src/lib389/lib389/schema.py

```diff
@@ -116,15 +116,51 @@ def get_attr_syntaxes(json=False):
             result = ATTR_SYNTAXES
         return result
 
-    def _get_schema_objects(self, object_model, json=False):
-        """Get all the schema objects for a specific model: Attribute, Objectclass,
-        or Matchingreule.
+    def gather_oc_sup_attrs(self, oc, sup_oc, ocs):
+        """
+        Recursively build up all the objectclass superiors' may/must
+        attributes
+        """
+        if sup_oc is None:
+            # First objectclass is none
+            sup_oc = oc
+        for soc in sup_oc['sup']:
+            if soc.lower() == "top":
+                continue
+            # Get sup_oc
+            for obj in ocs:
+                oc_dict = vars(ObjectClass(obj))
+                name = oc_dict['names'][0]
+                if name.lower() == soc.lower():
+                    # Found the superior, get it's attributes
+                    for attr in oc_dict['may']:
+                        if attr not in oc['may']:
+                            oc['may'] = oc['may'] + (attr,)
+                    for attr in oc_dict['must']:
+                        if attr not in oc['must']:
+                            oc['must'] = oc['must'] + (attr,)
+
+                    # Sort the tuples
+                    may = list(oc['may'])
+                    may.sort()
+                    oc['may'] = tuple(may)
+                    must = list(oc['must'])
+                    must.sort()
+                    oc['must'] = tuple(must)
+
+                    # Now recurse and check this objectclass
```

Comment:
I had a doubt about a possible infinite loop if for example OC1 -- sup --> OC2 -- sup --> OC3 -- sup --> OC1. But I was not able to find a testcase to produce this state (need to define superior before child and DEL/ADD superior if it has a child).

Reply:
Yeah this is possible with custom schema - can't trust an admin to be an LDAP expert.  I can keep track of OC's looked at, and break out if we see the same OC again

Reply:
Seems like the server will not allow me to create schema in this nested/looping fashion.

Reply:
I agree it'll be nice to add some `visited = set()` - just in case, as a safe precaution for a clean code.
Even though we see that the server guards it now.

Reply:
I have code to check it, but I seriously can not get the schema into a state where it can happen.  So I don't want to add that code as it's misleading since it's not possible.

Reply:
I get that it's not possible now, but adding the visited set is just good defensive programming - not misleading code at all, IMO.

If schema validation changes in the future or a bug slips in, we'd already be protected against the recursive loops. It's minimal code overhead for solid protection, and a pretty standard recursion pattern anyway.

Reply:
Fair enough, I'll add the code - it's not a big deal.  I guess I'm scarred by the fact that this product is so old that when you see the code doing some kind of check, you assume that it was there for a valid reason - that it was preventing a real problem.  That's why I used the word "misleading".  So I'll add some kind of comment clarifying that it's just being cautious.  I appreciate your persistence/intentions, and it is good practice as long as it's clear to the future reader.